src/components/Thumbs.js

```diff
@@ -26,81 +26,85 @@ class Thumbs extends Component {
         super(props);
 
         this.state = {
-            initialized: false,
             selectedItem: props.selectedItem,
             hasMount: false,
-            firstItem: this.getFirstItem(props.selectedItem),
-            images: []
+            firstItem: 0,
+            itemSize: null,
+            visibleItems: 0,
+            lastPosition: 0,
+            showArrows: false,
+            images: this.getImages()
         }
     }
 
-    componentDidMount (nextProps) {
-        if (!this.props.children) {
-            return;
-        }
-
+    componentDidMount(nextProps) {
         this.setupThumbs();
     }
 
-    componentWillReceiveProps (props, state) {
+    componentWillReceiveProps(props, state) {
         if (props.selectedItem !== this.state.selectedItem) {
             this.setState({
                 selectedItem: props.selectedItem,
                 firstItem: this.getFirstItem(props.selectedItem)
             });
         }
+        if (props.children !== this.props.children) {
+            this.setState({
+               images: this.getImages()
+            });
+        }
     }
 
     componentDidUpdate(prevProps) {
-        if (!prevProps.children && this.props.children && !this.state.initialized) {
-            this.setupThumbs();
+        if (this.props.children === prevProps.children) {
+            return;
         }
+
+        // This will capture any size changes for arrow adjustments etc.
+        // usually in the same render cycle so we don't see any flickers
+        this.updateSizes();
     }
 
     componentWillUnmount() {
         this.destroyThumbs();
     }
 
-    setupThumbs () {
+    setupThumbs() {
         // as the widths are calculated, we need to resize
         // the carousel when the window is resized
         window.addEventListener("resize", this.updateSizes);
         // issue #2 - image loading smaller
         window.addEventListener("DOMContentLoaded", this.updateSizes);
 
-        const images = this.getImages();
-
-        if (!images) {
-            return;
-        }
-
-        this.setState({
-            initialized: true,
-            images
-        });
-
         // when the component is rendered we need to calculate
         // the container size to adjust the responsive behaviour
         this.updateSizes();
     }
 
-    destroyThumbs () {
+    destroyThumbs() {
         // removing listeners
         window.removeEventListener("resize", this.updateSizes);
         window.removeEventListener("DOMContentLoaded", this.updateSizes);
     }
 
     updateSizes = () => {
-        if (!this.state.initialized) {
+        if (!this.props.children || !this.refs.itemsWrapper) {
```

Comment:
Maybe it clearer if this checks `this.state.images` but I thought it be OK and I'll await your thoughts on other things :smile: